Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+
+## Background Information
+
+### PowerShell Remoting
+
+PowerShell remoting consists of a number of internal engine components.
+It provides a way to establish a PowerShell session on a remote target, which can be another physical machine, virtual machine, or local process.
+Commands and scripts can then be run in the remote session, with data objects being returned to the client.
+PowerShell manipulates object instances and so returned data are objects that are serialized when transported over the remoting connection, and then deserialized after arriving at the destination.
+If the object type is known, PowerShell will deserialize it back into its original type.
+Unknown types are deserialized into a general PowerShell `PSCustomObject` type, that is essentially a property bag.  
+
+The core part of PowerShell remoting is the remoting protocol (PSRP) implementation, which handles all messages and data over the connection.
+But there is also the need to create the connections, along with mechanisms to transport protocol messages/data over the connection.
+These functions are considered to be outside PowerShell core remoting, and specific to each connection type.
+PowerShell provides abstractions for these functions, but leaves the implementation to each connection type.  
+
+#### Example - WinRM remoting
+
+This is the first remoting connection supported in PowerShell 2.0 and it is based on Windows WinRM.
+WinRM is Windows secure remote shell solution.
+PowerShell remoting uses WinRM to establish a remote connection, and then create a PowerShell session on a remote target.
+PowerShell uses the WinRM plugin architecture to create PowerShell target endpoint sessions as WinRM remote shell instances.
+The PowerShell engine provides WinRM specific connection objects, `WSManConnectionInfo`, to create a connection through WinRM.
+It also contains a transport implementation for transferring protocol messages and data through the WinRM connection.
+
+#### Example - SSH remoting
+
+SSH remoting was added to PowerShell v6.0 and is similar to WinRM remoting, in that SSH is also a secure remote shell solution.
+SSH remoting works on all PowerShell platforms (Linux, macOS, Windows), that have SSH installed.
+Similar to WinRM, SSH authenticates a user and establishes the PowerShell remoting connection to the target computer.
+The connection is defined in an SSH specific `SSHConnectionInfo` object.
+There is also an SSH specific transport implementation that transfer the protocol messages and data over the SSH connection.
+The target endpoint PowerShell session is created by SSH as a configured subsystem, where PowerShell runs in a server mode that handles remoting protocol messages.
+
+#### Example - Local process remoting
+
+PowerShell jobs (`Start-Job`) use PowerShell remoting but the remote connection is to a local child process instead of a remote machine.
```

Comment:
Is this done via named pipes or stdout/stdin? I would love to see how a named pipe session could be created.

Reply:
If you are talking about today you can do this already with https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.runspaces.namedpipeconnectioninfo?view=powershellsdk-7.0.0.

```powershell
$proc = Start-Process pwsh -PassThru
$proc.Id

# Since v6 this can be an actual name of a named pipe, otherwise pwsh calculates the
# expected name from the PID.
$connInfo = [System.Management.Automation.Runspaces.NamedPipeConnectionInfo]::new($proc.Id)
$rs = [RunspaceFactory]::CreateRunspace($connInfo)
$rs.Open()
$ps = [PowerShell]::Create()
$ps.Runspace = $rs
$ps.AddScript('$pid').Invoke()
$rs.Dispose()
```

I even have a testing script I use which uses named pipes to log the PSRP data exchanged which uses named pipe as communication https://github.com/jborean93/pypsrp/blob/asyncio/New-PSSessionLogger.ps1.

The trouble here is that there is no public way to create a PSSession with this mechanism, You can use reflection to create the PSSession object or just stick with the .NET API but that rules out using things like `Invoke-Command`. I'm hoping that this RFC makes it easier to 1 utilise named pipes in your own custom transports and 2 make it easy to use them with cmdlets that accept a PSSession object.

Reply:
`Start-Job` uses stdIn/stdOut.  However, there is a named pipe connection/transport implementation as well that you can look at.

https://github.com/PowerShell/PowerShell/blob/c2966acfd570a37076c28746db35a3bdcbd15412/src/System.Management.Automation/engine/remoting/fanin/OutOfProcTransportManager.cs#L1952